src/cnxml/se.tsx

```diff
@@ -0,0 +1,453 @@
+        let x = new XMLSerializer().serializeToString(document)
+
+        if (!x.startsWith('<?xml')) {
+            x = `<?xml version="1.0" encoding="${document.characterSet}">\n${x}`
+        }
+
+        return x
+    }
+}
+
+/** Serialization context */
+interface Context {
+    mediaMime: MediaMimeFunction
+}
+
+/** Attributes common to all elements */
+interface CommonAttrs {
+    /** Element's ID */
+    id: string
+    /** CSS classes */
+    class?: string
+}
+
+/** Serialize a single node */
+function serializeNode(editor: Editor, node: Node, ctx: Context): JSX.Node {
+    if (Text.isText(node)) {
+        return node.text
+    }
+
+    const children = Editor.hasInlines(editor, node)
+        ? serializeLine(editor, node, ctx)
+        : node.children.map(n => serializeNode(editor, n, ctx))
+
+    const attributes: CommonAttrs = {
+        id: node.id || `UUID${uuid.v4()}`,
+    }
+
+    if (WithClasses.hasClasses(node)) {
+        attributes.class = node.classes.join(' ')
+    }
+
+    for (const [test, serializer] of SERIALIZERS) {
+        if (test(node)) {
+            return serializer(node, attributes, children, ctx)
+        }
+    }
+
+    throw new Error(`no serializer defined for ${JSON.stringify(node)}`)
+}
+
+interface MarkedText extends Text {
+    emphasis?: boolean
+    strong?: boolean
+    position?: ''
+}
+
+/** Check if a node contains only plain, unmarked text */
+function isPlainText(node: Node): boolean {
+    if (Element.isElement(node)) {
+        return node.children.every(isPlainText)
+    }
+
+    const text = node as StyledText
+
+    return !text.emphasis
+        && !text.strong
+        && (text.position == null || text.position === 'normal')
+}
+
+/** Name of a style attribute on a {@link StyledText} node */
+type TextStyle = 'emphasis' | 'strong' | 'position'
+
+const STYLES: TextStyle[] = ['emphasis', 'strong', 'position']
+
+/** Map storing values of style attributes of a {@link StyledText} node */
+type Style = Map<TextStyle, string | boolean>
+
+/** Apply a single text style to a rendered inline node or text */
+function applyTextStyle(style: TextStyle, value: string | boolean, node: JSX.Node): JSX.Node {
+    switch (style) {
+    case 'emphasis':
+        return value ? <emphasis effect="italics">{node}</emphasis> : node
+
+    case 'strong':
+        return value ? <emphasis effect="bold">{node}</emphasis> : node
+
+    case 'position':
+        return value === 'superscript'
+            ? <sup>{node}</sup>
+            : value === 'subscript'
+                ? <sub>{node}</sub>
+                : node
+    }
+}
+
+/** Apply collection of text styles to a rendered inline node or text */
+function applyStyle(styles: Style, node: JSX.Node): JSX.Node {
+    for (const [style, value] of styles) {
+        node = applyTextStyle(style, value, node)
+    }
+
+    return node
+}
+
+/** Serialize a single line of text */
+function serializeLine(editor: Editor, node: Element, ctx: Context): JSX.Node {
```

Comment:
This function seems to be overcomplicated, I'll have to go back to it later and read it again